scipy/optimize/_hungarian.py

```diff
@@ -1,92 +1,105 @@
+       algorithms
+    2. Adjust weights so that each row has a zero entry. (Step 1 Wikipedia)
+    3. The computation of a maximal matching, done greedily. If the maximal
+       matching is maximum, we terminate (Step 2 Wikipedia)
+    4. Otherwise, we compute a minimum vertex cover of the bipartite graph
+       defined by zeros in the weighted bi-adjacency matrix.
+       (Step 3 in Wikipedia)
+    5. Subtract smallest unmarked element. Subtract from each other unmarked
+       element and add to every doubly marked element.  Return to step 2 above.
+       (Step 4 Wikipedia)
+    6. Once a result is found, an assignment (i.e. a maximum matching) is found
+       from the current known maximal matching by finding augmenting paths.
+
+Parameters
+     ----------
     cost_matrix : array
         The cost matrix of the bipartite graph.
-
-    Returns
-    -------
+    maximize : boolean
+        Specifies if the maximum weight matching should be computed. Default is
+        False, meaning that minimum weight matching is computed
+Returns
+ -------
     row_ind, col_ind : array
         An array of row indices and one of corresponding column indices giving
         the optimal assignment. The cost of the assignment can be computed
         as ``cost_matrix[row_ind, col_ind].sum()``. The row indices will be
         sorted; in the case of a square cost matrix they will be equal to
         ``numpy.arange(cost_matrix.shape[0])``.
 
-    Notes
-    -----
-    .. versionadded:: 0.17.0
-
-    Examples
-    --------
-    >>> cost = np.array([[4, 1, 3], [2, 0, 5], [3, 2, 2]])
-    >>> from scipy.optimize import linear_sum_assignment
-    >>> row_ind, col_ind = linear_sum_assignment(cost)
-    >>> col_ind
-    array([1, 0, 2])
-    >>> cost[row_ind, col_ind].sum()
-    5
-
     References
     ----------
-    1. http://csclab.murraystate.edu/bob.pilgrim/445/munkres.html
-
-    2. Harold W. Kuhn. The Hungarian Method for the assignment problem.
+    1. Harold W. Kuhn. The Hungarian Method for the assignment problem.
        *Naval Research Logistics Quarterly*, 2:83-97, 1955.
-
-    3. Harold W. Kuhn. Variants of the Hungarian method for assignment
+    2. Harold W. Kuhn. Variants of the Hungarian method for assignment
        problems. *Naval Research Logistics Quarterly*, 3: 253-258, 1956.
-
-    4. Munkres, J. Algorithms for the Assignment and Transportation Problems.
+    3. Munkres, J. Algorithms for the Assignment and Transportation Problems.
        *J. SIAM*, 5(1):32-38, March, 1957.
+    4. https://en.wikipedia.org/wiki/Hungarian_algorithm
+
+"""
+
+Node = namedtuple('Node', 'previous next_row next_col')
 
-    5. https://en.wikipedia.org/wiki/Hungarian_algorithm
+
+def linear_sum_assignment(cost_matrix, maximize=False):
+    """
+       Outward facing function. Provides validation and accesses internal
+        algorithm
     """
+
+    # This function should not have side effects on the cost matrix
+    cost_matrix = cost_matrix.copy()
```

Comment:
this isn't right if input is `array_like`, `copy` method isn't guaranteed to exist. Ideally you wouldn't need it (does your algorithm modify `cost_matrix`?), but if you do then write as `np.asarray(cost_matrix).copy()`